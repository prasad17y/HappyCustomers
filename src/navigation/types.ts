import {NavigationProp} from '@react-navigation/native';

export type RootStackParamList = {
  UserList: undefined;
  AddUser: undefined;
};

export type RootStackNavigationProp = NavigationProp<RootStackParamList>;
