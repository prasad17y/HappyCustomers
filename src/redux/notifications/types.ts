export type ToastType = 'success' | 'error' | 'info';

export interface ToastPayload {
  message: string;
  type: ToastType;
}

export interface NotificationsState {
  toast: ToastPayload | null;
}

export const SHOW_TOAST = 'SHOW_TOAST';
export const CLEAR_TOAST = 'CLEAR_TOAST';

interface ShowToastAction {
  type: typeof SHOW_TOAST;
  payload: ToastPayload;
}

interface ClearToastAction {
  type: typeof CLEAR_TOAST;
}

export type NotificationsActionTypes = ShowToastAction | ClearToastAction;
