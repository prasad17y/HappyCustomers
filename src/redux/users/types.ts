export interface UsersState {
  isSyncing: boolean;
  syncError: string | null;
  lastSyncTimestamp: string | null;
}

export const SYNC_USERS_REQUEST = 'SYNC_USERS_REQUEST';
export const SYNC_USERS_SUCCESS = 'SYNC_USERS_SUCCESS';
export const SYNC_USERS_FAILURE = 'SYNC_USERS_FAILURE';
export const CLEAR_SYNC_ERROR = 'CLEAR_SYNC_ERROR';

interface SyncUsersRequestAction {
  type: typeof SYNC_USERS_REQUEST;
  payload?: {forceRefresh: boolean};
}

interface SyncUsersSuccessAction {
  type: typeof SYNC_USERS_SUCCESS;
  payload: {didFetch: boolean};
}

interface SyncUsersFailureAction {
  type: typeof SYNC_USERS_FAILURE;
  payload: string;
}

interface ClearSyncErrorAction {
  type: typeof CLEAR_SYNC_ERROR;
}

export type UsersActionTypes =
  | SyncUsersRequestAction
  | SyncUsersSuccessAction
  | SyncUsersFailureAction
  | ClearSyncErrorAction;
