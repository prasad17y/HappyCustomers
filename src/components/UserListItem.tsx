import React from 'react';
import {View, Text, StyleSheet} from 'react-native';
import {UserType} from '../types';

interface UserListItemProps {
  user: UserType;
}

const UserListItem: React.FC<UserListItemProps> = ({user}) => {
  const getInitials = (name: string) => {
    const names = name.split(' ');
    const firstNameInitial = names[0] ? names[0][0] : '';
    return `${firstNameInitial}`.toUpperCase();
  };

  return (
    <View style={styles.container}>
      <View style={styles.initialsContainer}>
        <Text style={styles.initialsText}>{getInitials(user.name)}</Text>
      </View>
      <View style={styles.userInfo}>
        <Text style={styles.name}>{user.name}</Text>
      </View>
      {user.role === 'ADMIN' && <Text style={styles.role}>{user.role}</Text>}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 12,
    paddingHorizontal: 16,
    backgroundColor: '#fff',
    borderBottomWidth: 1,
    borderBottomColor: '#f0f0f0',
  },
  initialsContainer: {
    width: 40,
    height: 40,
    borderRadius: 4,
    backgroundColor: '#e3effa',
    justifyContent: 'center',
    alignItems: 'center',
    marginRight: 16,
  },
  initialsText: {
    color: '#0b5ac2',
    fontSize: 16,
    fontWeight: 'bold',
  },
  userInfo: {
    flex: 1,
  },
  name: {
    fontSize: 16,
    color: '#191a1a',
  },
  role: {
    fontSize: 14,
    color: '#888',
  },
});

export default React.memo(UserListItem);
