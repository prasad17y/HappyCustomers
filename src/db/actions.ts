import {Q} from '@nozbe/watermelondb';
import {database} from '.';
import client from '../apollo/client';
import {LIST_ZELLER_CUSTOMERS} from '../apollo/queries';
import UserModel from './models/UserModel';
import {Role} from '../types/types';

// representation of the raw JSON data we receive from the GraphQL API
interface ApiCustomer {
  id: string;
  name: string | null;
  email: string | null;
  role: string | null;
}

export const syncData = async () => {
  try {
    // fetch data from the grphQL API
    const {data} = await client.query({
      query: LIST_ZELLER_CUSTOMERS,
    });

    // to simulate network delay
    await new Promise(resolve => setTimeout(resolve, 1000));

    const customersFromAPI: ApiCustomer[] = data.listZellerCustomers.items;

    // Filter out any invalid records from the API
    const validCustomers = customersFromAPI.filter(customer => {
      if (
        !customer.name ||
        !customer.role ||
        !Object.values(Role).includes(customer.role as Role)
      ) {
        console.log(`Filtering out invalid user from API: ID ${customer.id}`);
        return false;
      }
      return true;
    });

    await database.write(async () => {
      const usersCollection = database.collections.get<UserModel>('users');

      // array of ids of validCustomers
      const validCustomerIds = validCustomers.map(c => c.id);

      // fetch all users from db that are also present in validCustomers
      const localUsersFound = await usersCollection
        .query(Q.where('id', Q.oneOf(validCustomerIds)))
        .fetch();

      const localUsersById = new Map(localUsersFound.map(u => [u.id, u]));

      const operations = validCustomers.map(customer => {
        const localUser = localUsersById.get(customer.id);

        // update if customer already exists
        if (localUser) {
          return localUser.prepareUpdate(user => {
            user.name = customer.name!;
            user.email = customer.email;
            user.role = customer.role! as Role;
          });
        }

        return usersCollection.prepareCreate(user => {
          user._raw.id = customer.id;
          user.name = customer.name!;
          user.email = customer.email;
          user.role = customer.role! as Role;
        });
      });

      await database.batch(...operations);
      console.log('Database synchronized successfully (Upsert complete)!');
    });
  } catch (error) {
    console.log('Failed to sync data:', error);
    throw error;
  }
};

interface AddUserParams {
  firstName: string;
  lastName: string;
  email: string | null;
  role: Role;
}

export const addUser = async ({
  firstName,
  lastName,
  email,
  role,
}: AddUserParams) => {
  try {
    await database.write(async () => {
      const usersCollection = database.collections.get<UserModel>('users');
      await usersCollection.create(user => {
        // random Id is generated automatically
        user.name = `${firstName} ${lastName}`;
        user.email = email;
        user.role = role;
      });
    });
    console.log('User added successfully!');
  } catch (error) {
    console.log('Failed to add user:', error);
    throw error;
  }
};
