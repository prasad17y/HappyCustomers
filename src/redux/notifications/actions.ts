import {
  SHOW_TOAST,
  CLEAR_TOAST,
  NotificationsActionTypes,
  ToastPayload,
} from './types';

export const showToast = (payload: ToastPayload): NotificationsActionTypes => ({
  type: SHOW_TOAST,
  payload,
});

export const clearToast = (): NotificationsActionTypes => ({
  type: CLEAR_TOAST,
});
