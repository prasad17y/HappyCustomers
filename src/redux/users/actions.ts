import {
  SYNC_USERS_REQUEST,
  SYNC_USERS_SUCCESS,
  SYNC_USERS_FAILURE,
  UsersActionTypes,
} from './types';

export const syncUsersRequest = (payload?: {
  forceRefresh: boolean;
}): UsersActionTypes => ({
  type: SYNC_USERS_REQUEST,
  payload,
});

export const syncUsersSuccess = (payload: {
  didFetch: boolean;
}): UsersActionTypes => ({
  type: SYNC_USERS_SUCCESS,
  payload,
});

export const syncUsersFailure = (error: string): UsersActionTypes => ({
  type: SYNC_USERS_FAILURE,
  payload: error,
});
