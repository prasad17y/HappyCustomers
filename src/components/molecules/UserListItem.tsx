import React from 'react';
import {View, Text, StyleSheet, TouchableOpacity} from 'react-native';
import {Role, UserType} from '../../types/types';
import Avatar from '../atoms/Avatar';
import Icon from '../atoms/Icon';
import {deleteUserRequest} from '../../redux/users/actions';
import {useNavigation} from '@react-navigation/native';
import {Screens} from '../../navigation/routes';
import {RootStackNavigationProp} from '../../navigation/types';
import {useAppDispatch} from '../../redux/hooks';

interface UserListItemProps {
  user: UserType;
}

const UserListItem: React.FC<UserListItemProps> = ({user}) => {
  const dispatch = useAppDispatch();
  const navigation = useNavigation<RootStackNavigationProp>();

  const handleItemPress = () => {
    navigation.navigate(Screens.EditUser, {userId: user.id});
  };

  const handleDeletePress = () => {
    dispatch(deleteUserRequest({userId: user.id, userName: user.name}));
  };

  return (
    <TouchableOpacity onPress={handleItemPress} style={styles.container}>
      <View style={styles.leftContainer}>
        <Avatar name={user.name} />
        <View style={styles.userInfo}>
          <Text style={styles.name}>{user.name}</Text>
        </View>
        {user.role === Role.ADMIN && <Text style={styles.role}>Admin</Text>}
      </View>
      <TouchableOpacity onPress={handleDeletePress} style={styles.deleteButton}>
        <Icon source={require('../../assets/trash.png')} />
      </TouchableOpacity>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingVertical: 12,
    paddingHorizontal: 16,
    backgroundColor: '#fff',
    borderBottomWidth: 1,
    borderBottomColor: '#f0f0f0',
  },
  leftContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
  },
  userInfo: {
    flex: 1,
    marginLeft: 16,
  },
  name: {
    fontSize: 16,
    color: '#191a1a',
  },
  role: {
    fontSize: 14,
    color: '#888',
    marginRight: 8,
  },
  deleteButton: {
    padding: 8,
  },
});

export default React.memo(UserListItem);
