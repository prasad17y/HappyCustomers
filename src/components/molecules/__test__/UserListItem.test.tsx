import React from 'react';
import {render, screen, fireEvent} from '@testing-library/react-native';
import UserListItem from '../UserListItem';
import {Role, UserType} from '../../../types/types';
import {deleteUserRequest} from '../../../redux/users/actions';

// Mock Redux's useDispatch hook
const mockDispatch = jest.fn();
jest.mock('../../../redux/hooks', () => ({
  useAppDispatch: () => mockDispatch,
}));

// Mock React Navigation's useNavigation hook
const mockNavigate = jest.fn();
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({
    navigate: mockNavigate,
  }),
}));

describe('UserListItem', () => {
  const mockUser: UserType = {
    id: '123',
    name: 'Jane Doe',
    email: 'jane.doe@example.com',
    role: Role.ADMIN,
  };

  beforeEach(() => {
    // Clear mock history before each test
    mockDispatch.mockClear();
    mockNavigate.mockClear();
  });

  it('should render user information correctly', () => {
    render(<UserListItem user={mockUser} />);
    expect(screen.getByText('Jane Doe')).toBeTruthy();
    expect(screen.getByText('Admin')).toBeTruthy();
  });

  it('should navigate to EditUser screen on press', () => {
    render(<UserListItem user={mockUser} />);

    fireEvent.press(screen.getByText('Jane Doe'));

    expect(mockNavigate).toHaveBeenCalledWith('EditUser', {userId: '123'});
  });

  it('should dispatch deleteUserRequest when the delete button is pressed', () => {
    render(<UserListItem user={mockUser} />);

    // Find the delete button by its testID and press it
    const deleteButton = screen.getByTestId('userlistitem-delete-button');
    fireEvent.press(deleteButton);

    // Verify that the correct Redux action was dispatched with the correct payload
    expect(mockDispatch).toHaveBeenCalledWith(
      deleteUserRequest({userId: '123', userName: 'Jane Doe'}),
    );
  });

  // snapshot test
  it('should match the snapshot', () => {
    const tree = render(<UserListItem user={mockUser} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
