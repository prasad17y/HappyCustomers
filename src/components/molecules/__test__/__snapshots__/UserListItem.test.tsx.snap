// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListItem should match the snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderBottomColor": "#f0f0f0",
      "borderBottomWidth": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#e3effa",
          "borderRadius": 4,
          "height": 40,
          "justifyContent": "center",
          "width": 40,
        }
      }
    >
      <Text
        style={
          {
            "color": "#0b5ac2",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        J
      </Text>
    </View>
    <View
      style={
        {
          "flex": 1,
          "marginLeft": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "#191a1a",
            "fontSize": 16,
          }
        }
      >
        Jane Doe
      </Text>
    </View>
    <Text
      style={
        {
          "color": "#888",
          "fontSize": 14,
          "marginRight": 8,
        }
      }
    >
      Admin
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
        "padding": 8,
      }
    }
    testID="userlistitem-delete-button"
  >
    <Image
      source={
        {
          "testUri": "../../../src/assets/trash.png",
        }
      }
      style={
        [
          {
            "height": 24,
            "width": 24,
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;
